### ISSUE TITLE
[bug] BazelDeps gets a binary instead of a library in `shared_library` parameter

### ISSUE THREAD

**Issue by marlamb** [conan-io/conan#15390]
### Environment details

* Ubuntu
* Conan version: Seen on `1.62.0`, but bug is also present on `develop2`
* Python version: 3.10

### Steps to reproduce

While experimenting with using `bazel` and build system and using `conan` as dependency manager I could detect that the `BUILD.bazel` files of some dependencies changed in consecutive `conan install` runs, although the installed packages did not change.

One example from the `conan-center-index` is the `protobuf` package, which has an executable `protoc` under `bin/protoc` and a library `libprotoc.so` under `lib/libprotoc.so`. Due to the `set` used for `libdirs` and `bindirs` in the [`_get_libs` function](https://github.com/conan-io/conan/blob/35a9d7d9e369662531e2a32d2893e1d7acb13ccc/conan/tools/google/bazeldeps.py#L112) the order in which `bindirs` and `libdirs` is iterated is not fixed, which causes sometimes the executable to be declared as a library in the resulting `BUILD.bazel` file.

I don't know what a good solution would be. I assume the `bindirs` are taken into account on purpose for use-cases with `libs` within these directories (if this is not the case, it would be my preferred solution not to include the `bindirs`). Also checking for naming collisions would be tricky, as can be seen by the `protobuf` example.

For reproduction of the problem if hope the following steps should work:
* create a recipe consuming `protobuf` and using the `BazelDeps` generator.
* run `conan install ...` multiple times.
* Check the `protobuf/BUILD.bazel` file for
   ```
   cc_import(
       name = "protoc_precompiled",
       shared_library = "lib/libprotoc.so" or "bin/protoc"
    )
   ```

Looking forward to your feedback and future solution!

**Comment by memsharded** [comment]
Thanks for reporting this @marlamb 

I am assigning it to @franramirez688 who implemented the integration.

I didn't fully understand what is the difference, I would only expect from the ``set()`` a possible change in order of things defined in the file, but not other issues. Maybe you could share two more complete files and how they differ? Thanks!

**Comment by marlamb** [comment]
Hi @memsharded , thanks for the quick reply.

I don't fully understand which files I could share, but I think I can explain the problem a little better. The problem is that the result is stored in a `dict` and both `protoc` and `libprotoc.so` result in the same key. So the iteration order determines the overall result -> the last one wins.

Lets look at the code simplified to the substantial part with some comments on what happens in the `protobuf` example for shared libraries on the Linux system.
```python
def _save_lib_path(lib_, lib_path_):
    lib_paths[lib_] = lib_path_
libdirs = cpp_info.libdirs
bindirs = cpp_info.bindirs
libs = cpp_info.libs[:]  # results in ["protoc"]
lib_paths = {}
for libdir in set(libdirs + bindirs):
    files = os.listdir(libdir)  # for "libdirs" contains "libprotoc.so" and for "bindirs" contains "protoc"
    for f in files:
        full_path = os.path.join(libdir, f)
        if f in libs:
            _save_lib_path(f, full_path)  # effect is lib_paths["protoc"] = "bin/protoc"
            continue
        name, ext = os.path.splitext(f)
        if name not in libs and name.startswith("lib"):
            name = name[3:]  # libprotoc -> protoc
        if ext in (".so", ".dylib", ".lib", ".dll")):
            if name in libs:
                _save_lib_path(name, full_path)  # effect is lib_paths["protoc"] = "lib/libprotoc.so"
                continue
```

Apart from that I now think the non-deterministic behavior might be only a symptom of two bigger problems:
* the generator in general does not handle name conflicts, such that multiple files potentially lead to only one entry in `lib_paths`.
* the generator potentially adds executables as shared libraries to a `cc_import` statement, which leads to the `bazel` error
  ```
  Error in fail: 'shared_library' does not produce any cc_import shared_library files (expected .so, .dylib or .dll)
  ```

Perhaps the title of this issue is therefore also not optimal. Feel free to adapt it.

I hope this helps :smiley: .

**Comment by franramirez688** [comment]
HI @marlamb 

Thank you for reporting this bug!

I think that https://github.com/conan-io/conan/pull/15410 should be enough to solve the issue. The problem was basically that the loop was checking directly the base name without seeing the extension so it was saving binaries too.

**Comment by franramirez688** [comment]
Solved by https://github.com/conan-io/conan/pull/15410

It'll be released in the next Conan 1.63 version. Thanks again!

**Comment by marlamb** [comment]
@franramirez688 Thanks for fixing the issue so quickly!
